File: boj/FloydWarshall/BOJ_9205.py
# 백준 9205 맥주 마시면서 걸어가기
# 21.03.01

import sys
import heapq
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
# sys.setrecursionlimit(10**6)

def getDist(a, b):
    return abs(pos[a][0] - pos[b][0]) + abs(pos[a][1] - pos[b][1])


INF = int(1e9)

t = int(input())

while t:
    t -= 1
    n = int(input())
    graph = [[INF] * (n+2) for _ in range(n+2)]
    pos = []
    for _ in range(n+2):
        x, y = map(int, sys.stdin.readline().split())
        pos.append((x, y))

    for i in range(n+2):
        for j in range(n+2):
            if i == j: continue
            elif getDist(i, j) <= 1000:
                graph[i][j] = getDist(i, j)

    for k in range(n+2):
        for i in range(n+2):
            for j in range(n+2):
                if i == j or i == k or j == k: continue
                graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])

    if graph[0][n+1] < INF: print("happy")
    else: print("sad")

File: boj/FloydWarshall/BOJ_9205.swift
// 백준 9250 맥주 마시면서 걸어가기
// 21.03.28

import Foundation

var t = Int(readLine()!)!
let INF = Int(1e9)

while t > 0 {
  t -= 1
  
  let n = Int(readLine()!)!
  var xy = [(Int, Int)]()
  var graph = [[Bool]](repeating: [Bool](repeating: false, count: n + 2), count: n + 2)
  
  for _ in 0..<n + 2 {
    let x = readLine()!.split(separator: " ").map { Int(String($0))! }
    
    xy.append((x[0], x[1]))
  }
  
  for i in stride(from: 0, to: n + 2, by: 1) {
    for j in stride(from: i + 1, to: n + 2, by: 1) {
      let dist = abs(xy[i].0 - xy[j].0) + abs(xy[i].1 - xy[j].1)
      if dist <= 1000 {
        graph[i][j] = true
        graph[j][i] = true
      }
    }
    
    for k in 1..<n + 1 {
      for i in 0..<n + 2{
        if !graph[i][k] { continue }
        for j in 0..<n + 2{
          if !graph[k][j] { continue }
          else { graph[i][j] = true }
        }
      }
    }
  }
  if graph[0][n+1] { print("happy") }
  else { print("sad") }
}
